### PR TITLE
Simplify k8s replicated set deployment

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
Simplify k8s replicated set deployment

## Motivation and Context
Continuation from #8629 which basically broke
zone deployments on k8s statefulset environment
due to incorrect assumptions which made it work
on a replicated set.

Fix this properly such that this container works
for both replicated set and stateful set deployment

## How to test this PR?
You need to test either with stateful set deployment or replicated set deployment.

```
~ helm install --set image.repository=y4m4/minio,image.tag=k8s,image.pullPolicy=Always,mode=distributed,persistence.enabled=false,replicas=4,zones=2 test stable/minio
```

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] Fixes a regression yes only in k8s environment #8629 
- [ ] Documentation needed
- [ ] Unit tests needed
- [ ] Functional tests needed (If yes, add [mint](https://github.com/minio/mint) PR # here: )

### CODE REVIEW

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :x: [more...](#9e761464-bf4c-4f07-a729-dc3cc023293c) |
| mint-large-bucket.sh | :x: [more...](#64f6d865-3804-4e2c-b7e8-6e520aedfe52) |
| mint-compression-fs.sh | :x: [more...](#f73e853a-9fe8-4008-befd-860b968ebda0) |
| mint-fs.sh | :x: [more...](#b6b0eb75-b9ed-4328-bf43-ba77e3392f47) |
<a name="64f6d865-3804-4e2c-b7e8-6e520aedfe52"></a>
### 8666-909b1a1/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.61:32229
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 0b5e52b576c6:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 2 minutes and 3 seconds
(2/13) Running aws-sdk-java tests ... done in 1 seconds
(3/13) Running aws-sdk-ruby tests ... done in 6 seconds
(4/13) Running mc tests ... done in 58 seconds
(5/13) Running minio-go tests ... FAILED in 3 minutes and 32 seconds
{
  "args": {},
  "duration": 105055,
  "error": "Please reduce your request",
  "function": "testStorageClassMetadataCopyObject()",
  "message": "CopyObject failed on RRS",
  "name": "minio-go: testStorageClassMetadataCopyObject",
  "status": "FAIL"
}

Executed 4 out of 13 tests successfully.
```
<a name="9e761464-bf4c-4f07-a729-dc3cc023293c"></a>
### 8666-909b1a1/mint-compression-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.51:32085
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 11d51ad9baaf:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 10 seconds
{
  "name": "awscli",
  "duration": 8253,
  "function": "make_bucket\n",
  "status": "FAIL",
  "error": "An error occurred (XMinioServerNotInitialized) when calling the CreateBucket operation (reached max retries: 4): Server not initialized, please try again."
}

Executed 0 out of 13 tests successfully.
```
<a name="f73e853a-9fe8-4008-befd-860b968ebda0"></a>
### 8666-909b1a1/mint-compression-fs.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.55:32331
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 250bab649081:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 5 minutes and 10 seconds
{
  "name": "awscli",
  "duration": 308727,
  "function": "make_bucket\n",
  "status": "FAIL",
  "error": "HTTPConnectionPool(host='72.28.97.55', port=32331): Max retries exceeded with url: /awscli-mint-test-bucket-16804 (Caused by ConnectTimeoutError(<botocore.awsrequest.AWSHTTPConnection object at 0x7fd7905090b8>, 'Connection to 72.28.97.55 timed out. (connect timeout=60)'))"
}

Executed 0 out of 13 tests successfully.
```
<a name="b6b0eb75-b9ed-4328-bf43-ba77e3392f47"></a>
### 8666-909b1a1/mint-fs.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.55:31378
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 98fef0745c4c:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 5 minutes and 16 seconds
{
  "name": "awscli",
  "duration": 314128,
  "function": "make_bucket\n",
  "status": "FAIL",
  "error": "HTTPConnectionPool(host='72.28.97.55', port=31378): Max retries exceeded with url: /awscli-mint-test-bucket-29098 (Caused by ConnectTimeoutError(<botocore.awsrequest.AWSHTTPConnection object at 0x7fb31a435128>, 'Connection to 72.28.97.55 timed out. (connect timeout=60)'))"
}

Executed 0 out of 13 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#094540f3-dc62-401b-aa91-20554d57067b) |
<a name="094540f3-dc62-401b-aa91-20554d57067b"></a>
### 8666-b9ac204/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.61:31133
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 25db346d5795:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 2 minutes and 12 seconds
(2/13) Running aws-sdk-java tests ... done in 1 seconds
(3/13) Running aws-sdk-ruby tests ... done in 6 seconds
(4/13) Running mc tests ... done in 1 minutes and 23 seconds
(5/13) Running minio-go tests ... FAILED in 3 minutes and 51 seconds
{
  "args": {},
  "duration": 104672,
  "error": "Please reduce your request",
  "function": "testStorageClassMetadataCopyObject()",
  "message": "CopyObject failed on RRS",
  "name": "minio-go: testStorageClassMetadataCopyObject",
  "status": "FAIL"
}

Executed 4 out of 13 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :x: [more...](#3c7de3c4-846f-4e50-bb5a-c4ced88a42da) |
| mint-xl.sh | :x: [more...](#42fb623b-9aa0-472f-8ef3-09f22a2e910a) |
| mint-large-bucket.sh | :x: [more...](#caf30499-a1fa-446f-98cf-b2e56328adb9) |
<a name="3c7de3c4-846f-4e50-bb5a-c4ced88a42da"></a>
### 8666-f377cbc/mint-compression-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.56:32029
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 68a9c4649e8d:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 10 seconds
{
  "name": "awscli",
  "duration": 8225,
  "function": "make_bucket\n",
  "status": "FAIL",
  "error": "An error occurred (XMinioServerNotInitialized) when calling the CreateBucket operation (reached max retries: 4): Server not initialized, please try again."
}

Executed 0 out of 13 tests successfully.
```
<a name="42fb623b-9aa0-472f-8ef3-09f22a2e910a"></a>
### 8666-f377cbc/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.52:30554
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp cf73f08b2634:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... FAILED in 6 seconds
{
  "name": "awscli",
  "duration": 4096,
  "function": "make_bucket\n",
  "status": "FAIL",
  "error": "An error occurred (XMinioServerNotInitialized) when calling the CreateBucket operation (reached max retries: 4): Server not initialized, please try again."
}

Executed 0 out of 13 tests successfully.
```
<a name="caf30499-a1fa-446f-98cf-b2e56328adb9"></a>
### 8666-f377cbc/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.51:30431
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 53765ce33846:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 2 minutes and 4 seconds
(2/13) Running aws-sdk-java tests ... done in 2 seconds
(3/13) Running aws-sdk-ruby tests ... done in 5 seconds
(4/13) Running mc tests ... done in 1 minutes and 14 seconds
(5/13) Running minio-go tests ... FAILED in 3 minutes and 33 seconds
{
  "args": {},
  "duration": 98293,
  "error": "Please reduce your request",
  "function": "testStorageClassMetadataCopyObject()",
  "message": "CopyObject failed on RRS",
  "name": "minio-go: testStorageClassMetadataCopyObject",
  "status": "FAIL"
}

Executed 4 out of 13 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Review by kannappanr — APPROVED** [review]
LGTM

File: cmd/endpoint.go [hunk 1]
```diff
@@ -301,8 +284,8 @@ func (endpoints Endpoints) UpdateIsLocal() error {
 					endpoints[i].Hostname(),
 				)
 
-				if orchestrated && endpoints.doAllHostsResolveToLocalhost() {
-					err := errors.New("hosts resolve to same IP, DNS not updated on k8s")
+				if orchestrated && endpoints.doAnyHostsResolveToLocalhost() {
+					err := errors.New("hosts resolve 127.*, DNS not updated on k8s")
```
Comment (krisis) [inline cmd/endpoint.go:288]: Should we remove k8s from the error message?

**Review by krisis — CHANGES_REQUESTED** [review]

File: cmd/endpoint.go (hunk 1 above)
Comment (harshavardhana) [reply cmd/endpoint.go:288]: It is on purpose since it is meant for k8s here @krisis

**Review by krisis — APPROVED** [review]
This error message can be fixed in a subsequent PR.

**Pull request merged**